File: DiaryNutrition/app/src/androidTest/java/ediger/diarynutrition/db/WeightDaoTest.java
package ediger.diarynutrition.db;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.List;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.room.Room;
import androidx.test.InstrumentationRegistry;
import androidx.test.runner.AndroidJUnit4;
import ediger.diarynutrition.LiveDataTestUtil;
import ediger.diarynutrition.data.source.DiaryDatabase;
import ediger.diarynutrition.data.source.entities.Weight;
import ediger.diarynutrition.data.source.dao.WeightDao;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.hamcrest.MatcherAssert.assertThat;

@RunWith(AndroidJUnit4.class)
public class WeightDaoTest {

    static final Weight WEIGHT = new Weight(75, 111111);

    @Rule
    public InstantTaskExecutorRule instantTaskExecutorRule = new InstantTaskExecutorRule();

    private DiaryDatabase mDatabase;
    private WeightDao mWeightDao;

    @Before
    public void initDb() {
        mDatabase = Room.inMemoryDatabaseBuilder(InstrumentationRegistry.getContext(),
                DiaryDatabase.class)
                .allowMainThreadQueries()
                .build();
        mWeightDao = mDatabase.weightDao();
    }

    @After
    public void closeDb() {
        mDatabase.close();
    }

    @Test
    public void insertAndGet() throws Exception {
        mWeightDao.insertWeight(WEIGHT);

        List<Weight> weights = LiveDataTestUtil.getValue(mWeightDao.getWeight(0, 200000));
        assertThat(weights.size(), is(1));
        assertWeight(weights.get(0), 1, WEIGHT.getAmount(), WEIGHT.getDatetime());
    }

    @Test
    public void insertAndLoadDesc() throws Exception {
        mWeightDao.insertWeight(new Weight(80, 100000));
        mWeightDao.insertWeight(WEIGHT);

        List<Weight> weights = LiveDataTestUtil.getValue(mWeightDao.loadWeight());
        assertThat(weights.size(), is(2));

        assertWeight(weights.get(0), 2, WEIGHT.getAmount(), WEIGHT.getDatetime());
    }

    @Test
    public void insertAndDeleteById() throws Exception {
        mWeightDao.insertWeight(WEIGHT);
        mWeightDao.deleteWeightById(1);

        List<Weight> weights = LiveDataTestUtil.getValue(mWeightDao.loadWeight());
        assertThat(weights.size(), is(0));
    }

    @Test
    public void insertAndUpdate() throws Exception {
        mWeightDao.insertWeight(WEIGHT);
        List<Weight> weights = LiveDataTestUtil.getValue(mWeightDao.getWeight(0, 200000));
        assertThat(weights.size(), is(1));

        Weight weight = weights.get(0);
        weight.setAmount(100);
        mWeightDao.updateWeight(weight);

        weights = LiveDataTestUtil.getValue(mWeightDao.getWeight(0, 200000));
        assertWeight(weights.get(0), 1, 100, WEIGHT.getDatetime());
    }

    private void assertWeight(Weight weight, int id, float amount, long datetime) {
        assertThat(weight, notNullValue());
        assertThat(weight.getId(), is(id));
        assertThat(weight.getAmount(), is(amount));
        assertThat(weight.getDatetime(), is(datetime));
    }

}

File: DiaryNutrition/app/src/androidTest/java/ediger/diarynutrition/db/WeightDaoTest.kt
package ediger.diarynutrition.db

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.room.Room
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import ediger.diarynutrition.data.source.DiaryDatabase
import ediger.diarynutrition.data.source.entities.Weight
import ediger.diarynutrition.getOrAwaitValue
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runBlockingTest
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.core.Is.`is`
import org.hamcrest.core.IsNull.notNullValue
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

fun Weight.assertWeight(original: Weight) {
    assertThat(this, notNullValue())
    assertThat(this.amount, `is`(original.amount))
    assertThat(this.datetime, `is`(original.datetime))
}

@ExperimentalCoroutinesApi
@RunWith(AndroidJUnit4::class)
class WeightDaoTest {

    @get:Rule
    var instantExecutorRule = InstantTaskExecutorRule()

    private lateinit var database: DiaryDatabase

    @Before
    fun initDb() = runBlockingTest {
        database = Room.inMemoryDatabaseBuilder(
                ApplicationProvider.getApplicationContext(),
                DiaryDatabase::class.java
        ).build()
    }

    @After
    fun closeDb() {
        database.close()
    }

    @Test
    fun insertAndGet() = runBlockingTest {
        database.weightDao().insertWeight(WEIGHT)

        val weightList = database.weightDao().getWeight()
        assertThat(weightList.size, `is`(1))

        weightList.first().assertWeight(WEIGHT)
    }

    @Test
    fun populateAndLoad() = runBlockingTest {
        val weight = Weight(80f, 222222)
        database.weightDao().populateWeight(listOf(WEIGHT, weight))

        val weightList = database.weightDao().loadWeight().getOrAwaitValue()
        assertThat(weightList.size, `is`(2))

        weightList.find { it.amount == WEIGHT.amount }!!.assertWeight(WEIGHT)
        weightList.find { it.amount == weight.amount }!!.assertWeight(weight)
    }

    @Test
    fun deleteById() = runBlockingTest {
        database.weightDao().insertWeight(WEIGHT)

        var weightList = database.weightDao().getWeight()
        assertThat(weightList.size, `is`(1))

        database.weightDao().deleteWeightById(1)

        weightList = database.weightDao().getWeight()
        assertThat(weightList.size, `is`(0))
    }

    @Test
    fun updateWeight() = runBlockingTest {
        database.weightDao().insertWeight(WEIGHT)

        val weight = database.weightDao().getWeight().first().apply {
            amount = 800f
        }

        database.weightDao().updateWeight(weight)

        database.weightDao().getWeight().first().apply {
            assertWeight(WEIGHT.apply { amount = 800f })
        }
    }

    companion object {
        val WEIGHT = Weight(75f, 111111)
    }
}